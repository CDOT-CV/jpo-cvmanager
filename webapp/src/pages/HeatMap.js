import React, { useEffect, useState } from 'react'
import MapGL, { Source, Layer, Popup } from 'react-map-gl'
import mbStyle from '../styles/mb_style.json'
import EnvironmentVars from '../EnvironmentVars'
import { useSelector, useDispatch } from 'react-redux'
import {
  selectLoading,
  selectRsuData,
  selectRsuCounts,
  selectSelectedRsu,
  selectMsgType,
  selectRsuIpv4,

  // actions
  selectRsu,
} from '../generalSlices/rsuSlice'

function HeatMap(props) {
  const dispatch = useDispatch()

  const loading = useSelector(selectLoading)

  const rsuData = useSelector(selectRsuData)
  const rsuCounts = useSelector(selectRsuCounts)
  const selectedRsu = useSelector(selectSelectedRsu)
  const msgType = useSelector(selectMsgType)
  const rsuIpv4 = useSelector(selectRsuIpv4)

  const [viewport, setViewport] = useState({
    latitude: 39.7392,
    longitude: -104.9903,
    width: '100%',
    height: props.auth ? 'calc(100vh - 135px)' : 'calc(100vh - 100px)',
    zoom: 10,
  })

  const [selectedRsuCount, setSelectedRsuCount] = useState(null)
  const [heatMapData, setHeatMapData] = useState({
    type: 'FeatureCollection',
    features: [],
  })

  useEffect(() => {
    const listener = (e) => {
      if (e.key === 'Escape') dispatch(selectRsu(null))
    }
    window.addEventListener('keydown', listener)

    return () => {
      window.removeEventListener('keydown', listener)
    }
  }, [dispatch])

  useEffect(() => {
    const localHeatMapData = { ...heatMapData }
    const heatMapFeatures = []
    rsuData.forEach((rsu) => {
      heatMapFeatures.push({
        type: 'Feature',
        geometry: {
          type: 'Point',
          coordinates: [rsu.geometry.coordinates[0], rsu.geometry.coordinates[1]],
        },
        properties: {
          ...rsu.properties,
          count: rsu.properties.ipv4_address in rsuCounts ? rsuCounts[rsu.properties.ipv4_address].count : 0,
        },
      })
    })
    localHeatMapData.features = heatMapFeatures
    setHeatMapData(localHeatMapData)
  }, [rsuData, rsuCounts])

  if (loading === false) {
  }

  function getStops() {
    // populate tmp array with rsuCounts to get max count value
    let max = Math.max(...Object.entries(rsuCounts).map(([, value]) => value.count))
    let stopsArray = [[0, 0.25]]
    let weight = 0.5
    for (let i = 1; i < max; i += 500) {
      stopsArray.push([i, weight])
      weight += 0.25
    }
    return stopsArray
  }

  const heatMapLayer = {
    id: 'heatmap',
    type: 'heatmap',
    maxzoom: 14,
    source: 'heatMapData',
    paint: {
      'heatmap-weight': {
        property: 'count',
        type: 'exponential',
        stops: getStops(),
      },
      'heatmap-intensity': ['interpolate', ['linear'], ['zoom'], 0, 0, 10, 1, 13, 2],
      'heatmap-color': [
        'interpolate',
        ['linear'],
        ['heatmap-density'],
        0,
        'rgba(33,102,172,0)',
        0.2,
        'rgb(103,169,207)',
        0.4,
        'rgb(209,229,240)',
        0.6,
        'rgb(253,219,199)',
        0.8,
        'rgb(239,138,98)',
        0.9,
        'rgb(255,201,101)',
      ],
      'heatmap-opacity': ['interpolate', ['linear'], ['zoom'], 10, 1, 13, 0.6, 14, 0],
    },
  }

  const rsuLayer = {
    id: 'rsuMarker',
    type: 'circle',
    source: 'heatMapData',
    minzoom: 13,
    paint: {
      'circle-radius': 8,
      'circle-color': 'rgb(255,201,101)',
      'circle-opacity': {
        stops: [[13, 1]],
      },
    },
  }

  return (
    <div className="container">
      <MapGL
        {...viewport}
        mapboxApiAccessToken={EnvironmentVars.MAPBOX_TOKEN}
        mapStyle={mbStyle}
        onClick={(e) => {
          if (e.features.length > 0 && e.features[0].layer !== undefined && e.features[0].layer.id === 'rsuMarker') {
            let rsu = {
              geometry: {
                coordinates: e.lngLat,
              },
              properties: e.features[0].properties,
            }
            dispatch(selectRsu(rsu))
            if (rsuCounts.hasOwnProperty(rsu.properties.ipv4_address))
              setSelectedRsuCount(rsuCounts[rsu.properties.ipv4_address].count)
            else setSelectedRsuCount(0)
          } else {
            console.debug('rsu not clicked')
          }
        }}
        onViewportChange={(viewport) => {
          setViewport(viewport)
        }}
      >
        <Source type="geojson" data={heatMapData}>
          <Layer {...heatMapLayer} />
          <Layer {...rsuLayer} />
        </Source>
        {selectedRsu ? (
          <Popup
            latitude={selectedRsu.geometry.coordinates[1]}
            longitude={selectedRsu.geometry.coordinates[0]}
            onClose={() => {
              dispatch(selectRsu(null))
              setSelectedRsuCount(null)
            }}
          >
            <div>
              <h2 className="popop-h2">{rsuIpv4}</h2>
              <p />
              <p className="popop-p">Milepost: {selectedRsu.properties.milepost}</p>
              <p className="popop-p">
                Serial Number: {selectedRsu.properties.serial_number ? selectedRsu.properties.serial_number : 'Unknown'}
              </p>
              <p className="popop-p">
                {msgType} Counts: {selectedRsuCount}
              </p>
            </div>
          </Popup>
        ) : null}
      </MapGL>
    </div>
  )
}

export default HeatMap
