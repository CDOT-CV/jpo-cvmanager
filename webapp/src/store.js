import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
