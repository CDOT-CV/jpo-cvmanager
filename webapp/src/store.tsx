import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import intersectionMapReducer from './components/map/map-slice'
import intersectionMapLayerStyleReducer from './components/map/map-layer-style-slice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
    },
    preloadedState,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
