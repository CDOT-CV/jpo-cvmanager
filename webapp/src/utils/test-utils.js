/**
 * This function searches for the every react-aria SSR ids in a given HTMLElement node and replace every attribute values with a static id
 *
 * This can be usefull when you're trying to generate a snapshot of components using react-aria under the hood
 *
 * @ex :
 * ```
 * const { container } = render(<Component />);
 *
 * replaceReactAriaIds(container);
 * ```
 *
 * @param container The HTMLElement node to search for SSR ids
 */
function replaceChaoticIds(container) {
  const props = [
    {
      selector: "class",
      updateFunc: (val) =>
        val.replace(
          /css-[0-9a-z]{6}-MuiTableCell-root-MuiTablePagination-root/g,
          "css-mocked-MuiTableCell-root-MuiTablePagination-root"
        ),
    },
  ];

  container.querySelectorAll("td").forEach((item) => {
    props.forEach((prop) => {
      if (item.getAttribute(prop.selector)) {
        item.setAttribute(prop.selector, prop.updateFunc(item.getAttribute(prop.selector)));
      }
    });
  });
  container.querySelectorAll("input[aria-invalid]").forEach((item) => {
    delete item["aria-invalid"];
  });
  return container;
}

export { replaceChaoticIds };
