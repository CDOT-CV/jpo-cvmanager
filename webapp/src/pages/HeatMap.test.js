import React from 'react'
import { render } from '@testing-library/react'
import HeatMap from './HeatMap'
import { Provider } from 'react-redux'
import { setupStore } from '../store'
import { replaceChaoticIds } from '../utils/test-utils'

it('should take a snapshot', () => {
  const { container } = render(
    <Provider store={setupStore({})}>
      <HeatMap />
    </Provider>
  )

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})
