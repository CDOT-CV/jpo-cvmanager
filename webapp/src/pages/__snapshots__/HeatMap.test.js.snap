// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      style="position: relative; width: 100%; height: calc(100vh - 100px); cursor: grab; outline: none;"
      tabindex="0"
    >
      <div
        style="position: relative; width: 100%; height: 100%;"
      >
        <div
          class=""
          style="position: absolute; width: 100%; height: 100%; overflow: hidden; visibility: inherit;"
        />
      </div>
    </div>
  </div>
</div>
`;
